design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
-    # Kubernetes resource). Passed to the function as the config block of its
-    # FunctionIO.
+  - name: compose-xr-using-go-templates
+    functionRef:
+      name: go-templates
     config:
-      apiVersion: database.example.org/v1alpha1
-      kind: Config
-      metadata:
-        name: cloudsql
-      spec:
-        version: POSTGRES_9_6
+      apiVersion: example.org/v1
+      kind: GoTemplate
+      source: Inline  # Or Remote, if you want to pull templates from git.
+      inline: |
+        {{- range $i := until ( . desired.composite.resource.spec.count ) }}
+        ---
+        apiVersion: iam.dummy.upbound.io/v1alpha1
+        kind: Robot
+        metadata:
+          annotations:
+            crossplane.io/external-name: {{ .desired.composite.resource | getExternalName }}-robot-{{ $i }}
+        spec:
+          forProvider:
+          {{- if $i | printf "robot-%d" | getComposedResource }}
+            # This resource already exists - use the same color.
+            color: {{ .spec.forProvider.color }}
+          {{ else }}
+            # The robot doesn't exist. Pick a new color.
+            color: {{ randomChoice "red" "green" "blue }}
+          {{- end }}
+        {{- end }}
 ```
 
-### Function API
+This example is functionally identical to [the Go example](#use-language-sdks),
+but uses a DSL - Go templates with http://masterminds.github.io/sprig/ and a
+few Crossplane-specific Functions.
 
-This document proposes that each function uses a `FunctionIO` type as its input
-and output. In the case of `Container` functions this would correspond to stdin
-and stdout. Crossplane would be responsible for reading stdout from the final
-function and applying its changes to the relevant XR and composed resources.
+Here's another example of a generic Function. This one lets Composition authors
+express their Composition logic in arbitrary, inline [Starlark][starlark].
 
 ```yaml
-apiVersion: apiextensions.crossplane.io/v1alpha1
-kind: FunctionIO
-config:
-  apiVersion: database.example.org/v1alpha1
-  kind: Config
-  metadata:
-    name: cloudsql
-  spec:
-    version: POSTGRES_9_6
-observed:
-  composite:
-    resource:
-      apiVersion: database.example.org/v1alpha1
-      kind: XPostgreSQLInstance
-      metadata:
-        name: my-db
-      spec:
-        parameters:
-          storageGB: 20
-        compositionSelector:
-          matchLabels:
-            provider: gcp
-      status:
-        conditions:
-        - type: Ready
-          status: True
-    connectionDetails:
-    - name: uri
-      value: postgresql://db.example.org:5432
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: compose-xr-using-go-templates
+    functionRef:
+      name: starlark
+    config:
+      apiVersion: example.org/v1
+      kind: StarlarkScript
+      script: |
```

Comment:
do you expect that all types of scripts/templates would have Remote option instead of Inline? This example doesn't imply Remote is a possibility, so I'm curious.

Reply:
These are just sketches of possible Functions, so it would really be up to the author of each Function.

In this case I figure the main utility is "I want to write a quick script inline of my Composition" so I left it out, but there's no reason you couldn't support Remote here too. As you pointed out on the Go Templates example the advantage of doing that is you get better syntax highlighting, linting, etc when your code isn't nested in an opaque string in a YAML document. 😄